kernel: k_pipe:  Rewrite k_pipe API

Mattemagikern opened the pull request:
The `k_pipe_*` API has been reworked to provide a more consistent and
intuitive interface. The new API aims to provide a simple to use byte
stream interface that is more in line with the POSIX pipe API.
The previous API has been deprecated and will be removed in a future
release.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
kernel: Rewrite k_pipe_* API

The `k_pipe_*` API has been reworked to provide a more consistent and
intuitive interface. The new API aims to provide a simple to use byte
stream interface that is more in line with the POSIX pipe API.
The previous API has been deprecated and will be removed in a future
release.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

andyross reviewed (commented):
This looks like a pretty solid path to me.  Some notes after a very quick read-through, the big ones being "fix the races" and "use ring_buf if possible".  You're also going to (obviously) need test coverage for all the cases here, including ideally a stress that will bang pipes with variant read/write/buffer sizes from multiple threads on each end.

The process and architecture people will want to chime in as to migration strategy.  As you'll discover I'm a lot (like... **a lot**) more tolerant of churn and cowboy evolution than most of the project.  My guess is that people are going to demand a more conservative rollout.  Maybe give the two APIs distinct names and let them live in the tree simultaneously for a version before flagging the old stuff deprecated.  Put each under kconfigs such that they can be independently en/disabled, etc...

Also recognize that treewide deprecation means combing through all the in-tree code to make sure that nothing else is using the old layers, deleting or filtering test cases using the wrong stuff, etc...  It's tedious and annoying and everyone will hate you when their precious out-of-tree code stops building; just be prepared. :)

Mattemagikern commented:
Thanks for your initial review  @andyross! 

I see an oportunity here I'd like to explore; you mention to use the `ring_buf` implementation instead of the one I have in this patch. I'm all for it, but when I'm already re-using: `k_sem`. 
If we can live with a small overhead in the `k_pipe` implementation we could piggyback on `k_sem` interface to have the scheduler management and locking mechanisms be managed. But perhaps the overhead is too much?

andyross commented:
That seems like an eminently reasonable choice too.  Though FWIW if you have a pipe based on pre-existing primitives then its should probably live in lib/ somewhere and not kernel/.    Whether the overhead is too high is sort of an app question.  As it happens k_sem is really pretty lean and optimal, and I'd guess that kind of tool would benchmark faster than the pipe we have already.  But surely you could do better at the bottom of the stack.

I love optimization, but don't know that we should demand it.

peter-mitsis commented:
In addition to the points that Andy raised, there are a couple of implications that are worth pointing out related to the copying of data.

1. Copying always goes from write buffer -> pipe buffer -> read buffer. One implication of this in the current proposal is that transfers can not exceed the size of the pipe buffer. This is not an unreasonable size limitation, but it should be documented.
2. Data completion and timeouts- part 1. There is a subtle change in how timeouts and data completion are being handled here as compared to the other data passing kernel objects. The others pend until the data is completely sent/received, while this new pipes implementation pend only until there is a chance that data can be sent/received.
3. Data completion and timeouts - part 2.  An indeterminate amount of time may elapse between the time at which the point the reader/writer is woken up and the time at which it executes to copy data into/out of the pipe buffer. This unknown amount of time may exceed the timeout. This subsequently raises the question ... should it copy the data? This could be further complicated should someone want to copy more data than will fit into the pipe buffer as both sender and receiver get swapped in and out to complete the whole transfer.

Mattemagikern commented on include/zephyr/kernel.h:
```diff
@@ -5175,10 +5185,133 @@ __syscall void k_pipe_buffer_flush(struct k_pipe *pipe);
+ * @param buffer Address of the pipe's buffer.
+ * @param buffer_size Size of the pipe's buffer.
+ */
+__syscall void k_pipe_init(struct k_pipe *pipe, uint8_t *buffer, size_t buffer_size);
```
TiL: cargo-cult - something is copied or replicated without a true understanding of its purpose or underlying principles.

I'm sitting here chuckling; I was caught, naturally.

Yeah, I have no clue how the `__syscall` / `__impl` works. I'll research this a bit and update the code once I've figured out how it works.

Mattemagikern replied:
I think this has been resolved now :+1:

Mattemagikern commented on include/zephyr/kernel.h:
```diff
@@ -5175,10 +5185,133 @@ __syscall void k_pipe_buffer_flush(struct k_pipe *pipe);
+	uint8_t flags;
+	size_t head;
+	size_t tail;
+	size_t count;
```
I looked into the `ring_buffer.c ` and it looks like it is two things in one? 
Both an object queue and a ring_buffer.c for bytes. 

The size of the ring_buffer struct alone would almost double the size of the k_pipe object so I opted to not use it. 
(Since I almost had all the logic in place for the ring-buffer already). 

And with your optimization: `head - tail % buffer_size` I've optimized away the count member, but ended up needing that space for the concurrency mechanism for the `flush`/`reset` function.

Mattemagikern commented:
@andyross 
> But surely you could do better at the bottom of the stack.

I’ve really enjoyed the learning experience of working with the scheduler, particularly with the `wait_q`, `pend`, and `unpend` functions. I’m also happy to put in the effort to make it as lean and efficient as possible. A small amount of effort here will pay dividends if the structure becomes useful to others.

@peter-mitsis 
1. Agreed, it should be documented. 
I started trying to answer 2. and 3. but my brain is fried atm. I'll sleep on the not so simple questions/concerns ^^

npitre reviewed (commented):
A few comments:

This was mentioned already, but you should consider using the ring_buffer
facility rather than open coding your own. This will save some binary
footprint not to have 2 different implementations linked in. Furthermore,
the most costly part in yours is by far the modulus operator on some targets
which the generic code dispense with.

You say:

> I looked into the `ring_buffer.c ` and it looks like it is two things in
> one? Both an object queue and a ring_buffer.c for bytes.

Not sure what you mean here, however if it is the `ring_buf_item_*` stuff
that looks confusing then you may safely ignore it. And better look in
`ring_buffer.h` first instead.

> The size of the ring_buffer struct alone would almost double the size of
> the k_pipe object so I opted to not use it.

Given those fields you will no longer need, the k_pipe object would grow by
only 16 bytes. Unless you are using a gazillon of pipes, the tradeoff is in
favor of code reuse both in terms of binary size and maintainability.

> (Since I almost had all the logic in place for the ring-buffer already).

Sorry but such arguments don't count.  ;-)

Next, the pipe implementation itself is lacking. If you want to write
10 bytes but there is room for only 5 bytes then you should write those 5 
bytes, flag any waiters, and wait up to the timeout to write the remaining 5
bytes. If suddenly you're woken up and there is only room for 3 bytes then 
you should write those 3 bytes, flag potential waiters, readjust the timeout
for what remains of the original one, and wait again for the last 2 bytes.
Same logic goes for the reader. Currently you return early when there is not
enough room despite the timeout not being reached.

Timeout readjusting is tricky. You may look at `sys_timepoint_timeout()` usage
in kheap.c for an example of how to do it simply.

Mattemagikern commented:
@peter-mitsis 

>2. Data completion and timeouts- part 1. There is a subtle change in how timeouts and data completion are being handled here as compared to the other data passing kernel objects. The others pend until the data is completely sent/received, while this new pipes implementation pend only until there is a chance that data can be sent/received.


The distinction you’ve pointed out between this implementation and other kernel objects is intentional. By waking up as soon as there’s a chance to send or receive data, it aims to maximize concurrency, reduce blocking, and decrease complexity. I understand that this behavior deviates from existing patterns, which might cause confusion or misaligned expectations. However, this new behavior aligns closely with the concept of a byte-stream, which I believe makes it more intuitive for certain use cases.

>  3. Data completion and timeouts - part 2. An indeterminate amount of time may elapse between the time at which the point the reader/writer is woken up and the time at which it executes to copy data into/out of the pipe buffer. This unknown amount of time may exceed the timeout. This subsequently raises the question ... should it copy the data? This could be further complicated should someone want to copy more data than will fit into the pipe buffer as both sender and receiver get swapped in and out to complete the whole transfer.

The issue of indeterminate time elapsing between wakeup and execution is indeed a valid concern, especially when considering scenarios where the timeout might expire during this window. The current design assumes that the timeout applies to the operation's initiation, not its completion. Achieving precise timing for both initiation and completion would be challenging. A possible middle ground might be to document the timeout behavior explicitly in the API to set clear expectations for users.

Regarding data transfers larger than the pipe buffer, this scenario is effectively handled by allowing operations to perform partial reads and writes. This mechanism ensures that even when the data size exceeds the buffer capacity, the transfer can proceed incrementally until completion.

Mattemagikern commented:
@npitre 

> Sorry but such arguments don't count. ;-)

Ahaha Fair! I’ve taken another look at the implementation and made revisions to utilize the ring_buffer as suggested. While I initially hesitated due to the additional overhead, I agree that the tradeoff in terms of binary size and maintainability justifies the change. Reusing existing infrastructure is the a better long-term choice.

Regarding the `ring_buffer_item_*` functions, I see how they could be perceived as adding unnecessary complexity, given their similarity to a message queue without synchronization. If they were refactored into their own type or removed, it might streamline the codebase further. However, this isn’t directly within the scope of this patch, so for now, I’ll focus on adapting the ring_buffer as it stands. :)

Mattemagikern commented:
Something that struck me when I made the move to `ring_buffer`: 

The `ring_buffer` implementation is not enabled by default. To use it, you need to set the Kconfig option: `CONFIG_RING_BUFFER`. Since kernel objects are expected to be enabled by default, I see the following options:

* Keep `k_pipe` in the kernel: Enable the `ring_buffer` implementation by default.
* Move `k_pipe` to `lib/`: Leave the `ring_buffer` configuration unchanged, with Kconfig being implicitly set.
* Keep `k_pipe` in the kernel: Implement the `ring_buffer` functionality directly within `k_pipe`, or create a shared `ring_buffer` for kernel objects, with k_msg_q and k_pipe as potential use cases.

I recommend keeping `k_pipe` in the kernel, as it is the only byte-stream implementation available. Depending on the optimization requirements, we can either:
* Keep the `ring_buffer` implementation, which would introduce 16 bytes of overhead to `k_pipe`, or
* Revert to the previous implementation and integrate the `ring_buffer` functionality directly into `k_pipe`, potentially duplicating a subset of the current `ring_buffer` logic.

What are your thoughts on this?

andyross commented:
There's already a CONFIG_PIPES that gets the pipe code, just have that "select RING_BUFFER".  It's true that our kconfigs are sometimes finer grained than they should be, but there's no shame in having the kernel depend on code in lib/ that's what it's there for.

npitre commented:
> What are your thoughts on this?

Many things rely on ring_buffer.c already, so I'm rather surprized it is 
not always compiled. I'd say it is light and quick to compile so no real
gain in leaving it out, given that bitarray.c is even bigger and always
built.

Mattemagikern commented:
I propose removing the `k_pipe_write_avail` and `k_pipe_read_avail` functions from the new `k_pipe` interface. These functions are problematic because their results are not guaranteed to remain valid, as the calling process does not hold a lock on the `k_pipe` struct. This lack of synchronization introduces potential race conditions, making the returned values unreliable in concurrent scenarios.

Including these functions in the interface suggests that their results can be relied upon for decision-making, which risks subtle bugs or misuse. If their intended purpose is to provide a "snapshot" for non-critical use cases, their utility seems limited.

For a more robust and intuitive interface, I recommend removing these functions entirely, as doing so would simplify the API and reduce potential confusion for users.

That said, these functions are currently used in `subsys/net/lib/sockets/socketpair.c` and `kernel/poll.c`. The socketpair implementation should naturally adapt to the new read/write functions. However, the implications for `poll.c` are less clear and may require further consideration.

Mattemagikern committed:
docs: update k_pipe documentation to align with the rewritten interface

This commit updates the k_pipe documentation to reflect the rewritten
interface.
Changes include:

* Detailed explanations of k_pipe concepts, initialization, and usage.
* Examples for defining, writing to, reading from, flushing,
	and closing a pipe.
* Clarified behavior in edge cases and introduced enhanced error-handling
  details.

Updated suggested use cases and API references.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
feat: enable polling support for k_pipe interface

This commit adds polling support to the newly rewritten k_pipe interface.
Changes include:

* Removed ifdef CONFIG_POLL from kernel/poll.c to let both implementations
  coexist.
* Added the needed datastructures to the new k_pipe struct.
* k_pipe_write(..) now notifies the poll subsystem that data is available.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
kernel: k_pipe: Add object_core support for the k_pipe rework

This patch adds object_core support for the k_pipe api rework.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
tests: kernel: pipe: Add new test cases for pipe API rework

This commit adds new test cases for the pipe API rework.
* basic.c: Sanity check for pipe operations.
* concurrency.c: Test pipe operations with multiple threads.
* stress.c: Test pipe operations under stress conditions.

And moves the old pipe test cases to the deprecated folder.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern commented on kernel/pipe.c:
```diff
@@ -0,0 +1,177 @@
+
+	__ASSERT_NO_MSG(pipe != NULL);
+	K_SPINLOCK(&pipe->lock) {
+		if (!(pipe->flags & PIPE_FLAG_OPEN)) {
```
I did change close due to flush already being defined as void and changing that api isn't an option :+1:

Mattemagikern commented on kernel/pipe.c:
```diff
@@ -0,0 +1,177 @@
+		pipe->tail = 0U;
+		pipe->count = 0U;
+
+		// or -ECONNRESET ?
```
I’ve made sure to document the behavior carefully to address potential ambiguity. 
In this case, the use of this specific error code is intended for a scenario where we want to signal to a consumer thread that the current work order or process is no longer valid, without needing to close the pipe or disrupt the threads that rely on it. I believe this is a valid use case that justifies going beyond the more common `-EINVAL` or `-EAGAIN`.

andyross commented on kernel/pipe.c:
```diff
@@ -0,0 +1,177 @@
+
+	__ASSERT_NO_MSG(((k_is_in_isr() == false) || K_TIMEOUT_EQ(timeout, K_NO_WAIT)));
+	__ASSERT_NO_MSG(pipe != NULL);
+	__ASSERT_NO_MSG(data != NULL);
```
I guess.  This doesn't rise to a -1, but we also want assertions to stay lean and easy, and Zephyr needs to build for a lot of surprisingly small systems (Nordic users say hi).  If we pollute the tree with a bunch of null checks, people are just going to stop building with CONFIG_ASSERT=y.

andyross commented on kernel/pipe.c:
```diff
@@ -0,0 +1,189 @@
+{
+	K_OOPS(K_SYSCALL_OBJ(pipe, K_OBJ_PIPE));
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(buffer, buffer_size));
+	K_OOPS(K_SYSCALL_MEMORY_READ(buffer, buffer_size));
```
You don't need both of these.  K_SYSCALL_MEMORY_WRITE checks for read permissions too.

andyross commented on kernel/pipes.c:
```diff
@@ -87,6 +88,16 @@ int z_impl_k_pipe_alloc_init(struct k_pipe *pipe, size_t size)
+{
+	K_OOPS(K_SYSCALL_OBJ_NEVER_INIT(pipe, K_OBJ_PIPE));
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(buffer, size));
+	K_OOPS(K_SYSCALL_MEMORY_READ(buffer, size));
```
Same, _WRITE is sufficient

andyross reviewed (commented):
Took another pass through.  Looking cleaner now.  The ring_buf usage looks correct and straightforward.  To repeat though: I'm absolutely not above appreciating a rewritten data structure, but if you do it you need to show clear and obvious advantages. :)

As others are pointing out, the semantics of short counts are a little confusing.  Zephyr IPC almost always takes timeout arguments, so ideally this code should sit in a loop trying to iteratively read/write until the timeout expires instead of returning on the first byte.

Code that actually does want to see the partial results (which is usually a rare edge case kind of thing, the whole idea of "file descriptor" style I/O is that it synthesizes a synchronous interface over async hardware behavior) can always try to read one byte as a sentinel and follow it with a call using K_NO_WAIT to grab the rest.

Mattemagikern commented:
Thanks, @andyross, @peter-mitsis , and @npitre for sharing your concerns about the semantics of short counts. I’ll revise the implementation to align with Zephyr IPC’s expected behavior and to finalize this PR with a tracing patch :+1: 
I plan to make these updates shortly after the holidays (starting the 25th). 

Thanks again for your reviews, and happy holidays!

Mattemagikern committed:
tracing: k_pipe: Add tracing support for reworked k_pipe API

Add tracing support for the reworked k_pipe API.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

npitre reviewed (commented):
some notes:

- `pipe_empty()` should probably use `ring_buf_is_empty()`.

- `sys_timepoint_timeout() is not free. I'd suggest moving it inside
  the "if pipe full/empty" conditionals.

- Documentation for `k_pipe_write()`says return value is 0 on success which is
  not right.

Mattemagikern committed:
drivers: ethernet: Rename struct ring_buf -> struct ring_buffer

The struct ring_buf is renamed to struct ring_buffer to be able to coexist
with the sys/ring_buffer.h header file.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
drivers: i2s: Rename struct ring_buf -> struct ring_buffer

The struct ring_buf is renamed to struct ring_buffer to be able to coexist
with the sys/ring_buffer.h header file.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
smp_shell: Add missing include for k_fifo

The smp_shell module uses k_fifo, but does not include the kernel header
file that defines it. This commit adds the missing include.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern commented:
> That said, these functions are currently used in subsys/net/lib/sockets/socketpair.c and kernel/poll.c. The socketpair implementation should naturally adapt to the new read/write functions. However, the implications for poll.c are less clear and may require further consideration.

I've looked into it a bit more and have found that the `k_pipe_write_avail` and `k_pipe_read_avail` may be needed after all. 
The `ioctl(..)` calls in `subsys/net/lib/sockets/socketpair.c`—such as `ZFD_IOCTL_FIONREAD` (and potentially `ZFD_IOCTL_POLL_UPDATE`) along with `spair_sendmsg()`, depend on the ability to determine how much data can be written to or read from the pipe. 

Should these function be implemented or do you have another idea?

npitre commented:
I guess `k_pipe_write_avail` and co can be implemented. They of course
would be reliable only if there is a single writer. The returned value 
might became inexact over time, but it could still be used as a minimum 
guaranteed value.

By the way, `k_pipe_write_avail()` is buggy as it may return `-eAGAIN` 
in two places, one with the lock un locked and the other with the lock 
still locked. This can be right, and the test suite ought to catch that.

npitre reviewed (commented):
And I meant "cannot be right" of course.

Mattemagikern commented:
Pardon me for the delayed response @npitre, had a dust with the flue. 

Regarding `k_pipe_write_avail` and `k_pipe_read_avail`: 
I appreciate the use case for these functions, especially as a minimum guaranteed value in single-writer scenarios. However, I find myself questioning whether they truly belong in the API. The `k_pipe` is designed as a primitive object to facilitate the producer/consumer pattern. Within this context, the concept of "availability" should generally be abstracted away, as the pipe's primary role is to handle the flow of data, not to inform users of transient states that may change immediately.

I’m concerned that providing these functions might encourage a reliance on snapshot-based views of pipe state, which could lead to race conditions or less efficient designs. However, implementing these functions is straightforward and would take minimal effort. 

As I see it there is two options: 
1. 
- Shift the unix socket implementation to use the ring-buffer instead. Since it is already protected by semaphores, this approach would ensure correctness without needing availability functions.
- This keeps the API minimalistic and focused on its core purpose, avoiding the potential misuse of these functions.
2. 
- Implement k_pipe_write_avail and k_pipe_read_avail but explicitly advise against their use in documentation. Highlight that their use indicates a potential design issue and that a message queue might be more appropriate in such cases.
- This option provides backward compatibility for the unix socket implementation and ensures developers are warned about the limitations and better alternatives.

I’m inclined towards Option 1 for the sake of API simplicity and to align with the intended use of k_pipe. However, I’d like to hear your thoughts, especially if there are compelling use cases where these functions might be indispensable.

@cfriedt : Since you’re familiar with the UNIX socket implementation, I’d greatly appreciate your guidance here. Do you see a strong case for keeping `k_pipe_write_avail` and `k_pipe_read_avail` given the context of UNIX sockets, or would a shift to a ring-buffer approach align better with the intended design?

cfriedt commented:
@Mattemagikern - IIRC, socketpair might need to use memory associated with a k_object of some kind in order to support userspace, e.g. a k_pipe.

You can always try to add a commit that removes the k_pipe, and then run the testsuite.

Mattemagikern committed:
net: socketpair to use ring_buffer instead of k_pipe

Replaced the k_pipe-based implementation in sockpair with ring_buffer
based implementation instead.
The move to ring_buffer is done to avoid overhead of k_pipe and to align
with the new k_pipe API.
This does not pose any added risk to concurrency as the read and write
functions are protected by semaphores for both spairs.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
tests: Update tests to use new k_pipe API

Update tests to use the reworked k_pipe API.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern commented:
Twister 1 (`frdm_k64f_mk64f12/samples/subsys/tracing/sample.tracing.percepio`) fails due to the trace
definition in `modules/debug/percepio` being out of sync with the Zephyr kernel, I think. Since it
is undefined prior to being redefined, I think there is some sort of error here.

The following modification to the trace mappings resolves the issue for me:
```
diff --git a/TraceRecorder/kernelports/Zephyr/include/tracing_tracerecorder.h b/TraceRecorder/kernelports/Zephyr/include/tracing_tracerecorder.h
index 75e7cdf..42b0d81 100644
--- a/TraceRecorder/kernelports/Zephyr/include/tracing_tracerecorder.h
+++ b/TraceRecorder/kernelports/Zephyr/include/tracing_tracerecorder.h
@@ -681,7 +681,7 @@ extern "C" {

 /* Pipe trace mappings */
 #undef sys_port_trace_k_pipe_init
-#define sys_port_trace_k_pipe_init(pipe, ...)                                       \
+#define sys_port_trace_k_pipe_init(pipe, buffer, size)                              \
     sys_trace_k_pipe_init(pipe, buffer, size)
 #undef sys_port_trace_k_pipe_cleanup_enter
 #define sys_port_trace_k_pipe_cleanup_enter(pipe, ...)
```

Twister 2, seems to fail due to timeout rellated issues both in the `pipe_api` tests and some dhcpv6
tests. Assistance or insights into these tests would be appreciated, I'm not sure what the issue is.

npitre commented:
> - Shift the unix socket implementation to use the ring-buffer instead. Since it is already protected by semaphores, this approach would ensure correctness without needing availability functions.

OK. I do like that. Makes that code much more efficient as well as it 
actually does not need the overhead of a full pipe facility 
(doesn't want thread suspension).

npitre reviewed (commented):
Looks great.

Some more details: I'd suggest removing the `wait_cond_t` abstraction
and simply deal with the pipe full/empty test at the `wait_for()` call 
site instead. Doing so will let the compiler inline them which is likely 
to produce smaller and faster code compared to passing function pointers.

And with regards to tracing: you have many exit points invoking the same
trace facility. Might be worth consolidating them with something like:

```
                [...]
                if (rc == -EAGAIN) {
                        k_spin_unlock(&pipe->lock, key);
                        rc = read ? read : -EAGAIN;
                        goto exit;
                } [...]

        }
        rc = read;

exit:
        SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_pipe, read, pipe, rc);
        return rc;
}
```

andyross commented on subsys/net/lib/sockets/socketpair.c:
```diff
@@ -550,10 +550,7 @@ static ssize_t spair_write(void *obj, const void *buffer, size_t count)
-			 &bytes_written, 1, K_NO_WAIT);
-	__ASSERT(res == 0, "k_pipe_put() failed: %d", res);
-
+	bytes_written = ring_buf_put(&remote->recv_q, (void *)buffer, count);
```
I know it's claimed so in a comment, but can you be explicit about the synchronization in the commit message.  Basically: k_pipe is a threadsafe synchronized tool, while ring_buf is a raw data structure and isn't.  I just want to see a statement to the effect of "All the locations being changed here are already protected by this spinlock (or semaphore, or whatever)" and don't need further locking.

Mattemagikern replied:
I updated the commit message accordingly :+1:

andyross reviewed (dismissed):
I think this looks good for me at this point, with one nitpick about being 100% sure we audited the synchronization in socketpair.

peter-mitsis commented on kernel/pipe.c:
```diff
@@ -0,0 +1,226 @@
+{
+	struct k_thread *thread_to_unblock = z_unpend_first_thread(waitq);
+
+	if (likely(thread_to_unblock)) {
```
thread_to_unblock -> thread_to_unblock != NULL

This is from the coding guidelines, Zephyr Rule 85.

> The controlling expression of an if statement and the controlling expression of an iteration-statement shall have essentially Boolean type

Essentially, unless the type being tested is already a boolean, perform an explicit comparison to something of the same type.

This applies elsewhere as well.

peter-mitsis commented on kernel/pipe.c:
```diff
@@ -0,0 +1,226 @@
+			pipe->flags &= ~PIPE_FLAG_RESET;
+		}
+		return -ECANCELED;
+	} else if (!pipe_full(pipe)) {
```
There has been something nagging me about this and how this integrates with the caller, and I think I have finally figured out what it is. Please correct me if I am wrong, or if I have misunderstood something. (Also, perhaps this has been covered in another commit, but so far I have been just looking at this one).

There does not appear to be any differentiation between a timeout and "stolen space". Imagine that thread A wants to write to the pipe, but it pends because it is full. Thread B reads from the pipe and wakes thread A to write more data into it. However, before thread A can do so, thread C comes along at a higher priority and writes to the pipe before thread A gets a chance to do so. If thread A executes before thread B, then thread A may find that the pipe is full, and it will return -EAGAIN. This can cause the caller (k_pipe_write) to return prematurely before the timeout has expired.

In some scenarios, this will result in k_pipe_write() returning a value indicating that some bytes have been written, while in others it will result in k_pipe_write() returning -EAGAIN even though the timeout may still be far off into the future. This is in contradiction to the behavior stated in the comment header for k_pipe_write().

I suspect something similar may happen with a thread trying to read data from the pipe.

Mattemagikern replied:
> There does not appear to be any differentiation between a timeout and "stolen space". Imagine that thread A wants to write to the pipe, but it pends because it is full. Thread B reads from the pipe and wakes thread A to write more data into it. However, before thread A can do so, thread C comes along at a higher priority and writes to the pipe before thread A gets a chance to do so. If thread A executes before thread B, then thread A may find that the pipe is full, and it will return -EAGAIN. This can cause the caller (k_pipe_write) to return prematurely before the timeout has expired.

Very true, great catch! I hadn't considered that scenario before, but now that you mention it, It got me thinking of another case: 
Let’s say `Thread A` writes only chunks of `0xFF`, `Thread B` writes only chunks of `0xAA,` and `Thread C `is reading from the pipe. In this scenario, there’s no guarantee that full chunks of data from a single thread will be delivered as written. The only assurance is that the maximum amount of data that can be delivered within the timeout will be transferred.

Additionally, the sequence of data read by Thread C could be arbitrary and depend on thread scheduling. This means that Thread C might receive interleaved data from Thread A and Thread B in an order that doesn’t preserve any logical grouping or sequence, especially under high contention or preemption conditions. But somehow it aligns with my expectations of a pipe, it just accepts bytes and funnles it to anether thread. The race-condition is already introduced by the multiple producers and message boundries concept.

Regardless, I will resolve the issue raised regarding stolen-space asap!

nashif commented:
> Twister 2 tests seems to fail on a specific target `m2gl025_miv/miv` which I don't have access to, and there seems to be a timeout issue of the stress-tests, but since they work on other targets I'm not sure what the issue is here. Do anyone of you have any insigts here?

this is using renode, you can install it and run it as a simulator like qemu.

npitre reviewed (commented):
Still an issue.

You have:

```
static inline int wait_for_data(_wait_q_t *waitq, struct k_pipe *pipe, k_spinlock_key_t *key,
                                k_timeout_t timeout)
{
        [...]
        } else if (!pipe_empty(pipe)) {
                return 0;
        }
        return -EAGAIN;
}
```

Let's assume this is called by `k_pipe_write()` because the pipe is full.
The thread goes to sleep until a reader has consumed data from the pipe.
Now if the pipe is completely empty you return `-EAGAIN` even if there is
plenty of space to write data? This looks like a bug that the test suite
should catch.

I'd suggest moving the last test to `k_pipe_read()` since this is the only
caller that cares about such condition and have the final return to be 0.

nashif commented on include/zephyr/kernel.h:
```diff
@@ -5101,9 +5113,10 @@ int k_pipe_cleanup(struct k_pipe *pipe);
+__deprecated __syscall int k_pipe_alloc_init(struct k_pipe *pipe, size_t size);
 
 /**
+ * @deprecated
```
please add the alternative and what replaces the deprecated function here

nashif commented on kernel/Kconfig:
```diff
@@ -713,6 +714,7 @@ config EVENTS
 
 config PIPES
 	bool "Pipe objects"
+	select DEPRECATED
```
This might be seen as a deprecation of pipes, what you are doing here is deprecating the kconfig only, worth some text in the help section.

nashif commented on kernel/Kconfig.obj_core:
```diff
@@ -77,7 +77,7 @@ config OBJ_CORE_SEM
 config OBJ_CORE_PIPE
 	bool "Integrate pipe into object core framework"
-	default y if PIPES
+	default y
```
So pipes are now enabled by default, this might have some footprint implication on those not using the pipe, why does it need to be enabled by default?

Mattemagikern replied:
> why does it need to be enabled by default?

Strictly speaking it deosn't **need** to be. However, similar data structures are typically enabled automatically when `CONFIG_MULTITHREADING` is set. In this case, the pipe implementation and atomic services are the only data structures treated differently.
I assumed enabling it by default would have minimal impact, as the additional pipe structure would only slightly increase build time and should be optimized away during compilation.

Moreover, this approach eliminates the need for an extra Kconfig option to manage mutually exclusive configurations when selecting a pipe implementation during the deprecation process.

Do you prefer an alternative approach to handle this?

npitre reviewed (commented):
Ok... I need to revise my previous comment as I didn't realize there are
separate `wait_for_data()` and `wait_for_space()` now.

The obvious question is... why?

Those two functions are identical except for the last condition. Hence
my earlier suggestion to remove that last condition, return 0 at the end,
and perform that last condition in the caller instead

Mattemagikern commented:
Thank you @peter-mitsis, @npitre, @andyross, and @nashif for your reviews! 

I think I've made the appropriate changes based on your feedback. 

The PR that resolves the Twister 1: https://github.com/zephyrproject-rtos/percepio/pull/14

andyross reviewed (dismissed):
Driving by to refresh +1.  One naming nitpick I just noticed, but doesn't rise to a respin if no one else cares.

Mattemagikern commented on include/zephyr/kernel.h:
```diff
@@ -4991,6 +4992,28 @@ void k_mbox_data_get(struct k_mbox_msg *rx_msg, void *buffer);
+ *
+ * @param pipe Address of the pipe.
+ */
+__syscall void k_pipe_flush(struct k_pipe *pipe);
```
I agree, and I found a bug when I revisited the flush/reset operation so I need to make a patch regardless. 

I'll sort this out after my workday :+1:

cfriedt commented:
I'm a bit confused about the `twister` failure with 
```shell
west twister -p frdm_k64f/mk64f12 -s sample.tracing.percepio --no-detailed-test-id
```
The build below works fine
```shell
west build -p auto -b frdm_k64f/mk64f12 samples/subsys/tracing -- -DCONFIG_COMPILER_WARNINGS_AS_ERRORS=y
```

Manually ran 
```shell
west build -p auto -b qemu_x86 -t run tests/net/socket/socketpair
west build -p auto -b qemu_x86 -t run tests/posix/net
```

and they are both fine.

Mattemagikern commented:
@cfriedt 

> but it should probably be green before this gets merged.

Agreed! I reached out to @eriktamlin on discord regarding the patch in [modules/debug/percepio](https://github.com/zephyrproject-rtos/percepio/pull/14). 
I realized that the patch that resolves the build issue in `Twister 1` will mask the issue, and the functionallity the module provides will be voided for the new implementation. 

I'd like to make you all aware of the latest changes where I renamed the `k_pipe_flush(...)` => `k_pipe_reset(...)` and I've fixed a bug regarding reading/writing to a pipe while it is resetting. 

The case is this: 
Thread A waits for data, Thread B calls the `k_pipe_reset(...)`, Thread C tries to 
[read|write] [from|to] the pipe before Thread A has exited and flipped the `K_PIPE_RESET` flag. 
Previously Thread C would exit with `-EAGAIN`, now it exits with `-ECANCELED` as intended, an unlikely edge case but it should be fixed now.

cfriedt commented:
@Mattemagikern - it might make sense to create a bug report and associate it with your PR. That may get the attention of the maintainer faster.

peter-mitsis reviewed (dismissed):
I think this works!

npitre reviewed (commented):
Well... one more thing.

Many `zassert_true()` message strings are wrong. They look like bad
copy-paste. For correctness sake those should be fixed.

Alternatively, you could get rid of them entirely. Empty strings are still
better than wrong strings IMHO.

Mattemagikern commented:
@npitre Agreed, It was sloppy of me. 
I think the new revision is more clear! :)

cfriedt commented on tests/kernel/pipe/deprecated/pipe_api/src/test_pipe_fail.c:
```diff
@@ -3,8 +3,6 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
-
```
@Mattemagikern - can you please drop unnecessary whitespace changes?

Mattemagikern replied:
Certainly! 

I think I got them all! :)

Mattemagikern commented:
Good evening, everyone!

Thanks to @eriktamlin at Percepio, the integration tests for the `modules/debug/percepio` trace functionality should now pass. Unless something unexpected arises during this test run, I anticipate all tests will complete successfully.

I believe all feedback has been addressed at this stage. I’m switching the PR from Draft to Ready for Review. Please let me know if there’s anything further that needs attention.

Thank you for your time and input!

kartben commented:
@Mattemagikern please update https://github.com/zephyrproject-rtos/zephyr/issues/80539 to track deprecation

peter-mitsis commented:
I approved too soon I think. I realized something after reviewing the extension to this by @npitre (see #84052). At some point after the waiter is notified, we need to invoke _z_reschedule()_ to force a schedule point. Otherwise we are relying upon some external action to force a reschedule (such as an interrupt, pending on an object, ...)

npitre commented:
I'll add the `z_reschedule()` to my follo-up patches.
